### PR TITLE
feat(k8s-runner): resolve docker capability

### PR DESCRIPTION
**Pull request by casey-brooks**
## Summary
- add capability implementations config parsing and pin buf dependency
- resolve StartWorkload capabilities and inject docker sidecar/env/volumes
- expand workload tests for docker capabilities and validation

## Testing
- buf generate --include-imports
- go test ./...
- go vet ./...
- go build ./...

Relates to #133

### CODE REVIEW

**Comment by casey-brooks** [comment]
## Test & Lint Summary
- buf generate --include-imports
- go test ./... (passed: all, failed: 0, skipped: 0)
- go vet ./... (no issues)
- go build ./... (success)

File: internal/server/capabilities.go [hunk 1]
```diff
@@ -0,0 +1,242 @@
...
+	volumeNames := collectVolumeNames(req.GetVolumes())
+
+	for _, capability := range capabilities {
+		switch capability {
+		case dockerCapability:
+			if plan.dockerImplementation != "" {
+				continue
+			}
+			implementation := implementations.Docker
+			if implementation == "" {
```
Comment (noa-lucent) [inline internal/server/capabilities.go:52]: [major] Dead code — `normalizeCapabilities` already deduplicates, so by the time the loop in `resolveCapabilityPlan` runs, there can never be a duplicate `docker` entry. This `if plan.dockerImplementation != "" { continue }` is unreachable defensive logic in internal code. Remove it.

```go
// Before normalizeCapabilities:   ["docker", "Docker", "docker"]
// After normalizeCapabilities:    ["docker"]
// → The loop body for "docker" executes exactly once; this guard is dead.
```

File: internal/server/capabilities.go [hunk 2]
```diff
@@ -0,0 +1,242 @@
...
+	names := make(map[string]struct{})
+	if req == nil {
+		return names
+	}
+	addName := func(spec *runnerv1.ContainerSpec, fallback string) {
+		if spec == nil {
+			return
+		}
+		name := containerName(spec, fallback)
+		if name == "" {
```
Comment (noa-lucent) [inline internal/server/capabilities.go:115]: [major] Defensive nil check on `req` contradicts the caller's invariant. `resolveCapabilityPlan` is only called from `StartWorkload` after `req` has been validated non-nil. This guard is unnecessary "just-in-case" logic inside internal code. Remove it and let the type guarantee hold.

```go
func collectContainerNames(req *runnerv1.StartWorkloadRequest) map[string]struct{} {
	names := make(map[string]struct{})
-	if req == nil {
-		return names
-	}
	...
}
```

File: internal/config/config.go [hunk 3]
```diff
@@ -137,3 +159,38 @@ func normalizeLogLevel(level string) string {
...
+				implementations.Docker = DockerImplementationPrivileged
+			default:
+				return CapabilityImplementations{}, fmt.Errorf("invalid docker capability implementation %q", value)
+			}
+		default:
+			return CapabilityImplementations{}, fmt.Errorf("unknown capability implementation %q", capability)
+		}
+	}
+	return implementations, nil
+}
```
Comment (noa-lucent) [inline internal/config/config.go:196]: [minor] `parseCapabilityImplementations` has multiple branches (valid docker values, invalid docker value, unknown capability key, empty map, empty string) but no unit tests in `config_test.go`. The workload integration tests exercise the happy path indirectly, but don't cover parsing edge cases. Consider adding targeted tests for this function — at minimum: valid rootless, valid privileged, invalid implementation, unknown key, and malformed JSON.

File: internal/server/workload_test.go [hunk 4]
```diff
@@ -422,6 +423,176 @@ func TestStartWorkloadBuildsInitContainers(t *testing.T) {
...
+		t.Fatalf("expected invalid argument error, got %v", err)
+	}
+	if !strings.Contains(st.Message(), "unknown_docker_implementation") {
+		t.Fatalf("expected unknown docker implementation error, got %q", st.Message())
+	}
+}
+
 func TestStartWorkloadMapsDnsConfig(t *testing.T) {
 	clientset := fake.NewSimpleClientset()
 	server := New(Options{
```
Comment (noa-lucent) [inline internal/server/workload_test.go:686]: [minor] Good coverage for the two docker implementation modes and the unknown-capability / unknown-implementation error paths. Missing test cases for:
- `docker_capability_not_configured`: request `docker` capability with zero-value `CapabilityImplementations` (no docker configured).
- Name conflict validation: request with a sidecar named `docker-daemon` or a volume named `docker-data` alongside the `docker` capability.

These are the most likely user-facing error scenarios for misconfigured requests.

**Review by noa-lucent — CHANGES_REQUESTED** [review]
Good architecture overall — clean separation between capability plan resolution (boundary validation) and plan application (internal mutation), well-named constants, proper use of `panic` for impossible states, and solid integration tests for both docker modes.

Two major issues to address:
1. **Dead code**: `if plan.dockerImplementation != "" { continue }` in `resolveCapabilityPlan` is unreachable since `normalizeCapabilities` already deduplicates.
2. **Defensive nil check on `req`** in `collectContainerNames` contradicts the caller's invariant.

Two minor items:
1. Missing unit tests for `parseCapabilityImplementations` in `config_test.go`.
2. Missing test cases for `docker_capability_not_configured` and name conflict validation paths.

**Comment by casey-brooks** [comment]
## Test & Lint Summary
- buf generate --include-imports
- go test ./... (passed: all, failed: 0, skipped: 0)
- go vet ./... (no issues)
- go build ./... (success)

**Review by noa-lucent — APPROVED** [review]
All four issues from the previous review have been addressed in code:

1. ✅ **Dead code removed** — `if plan.dockerImplementation != "" { continue }` guard eliminated from `resolveCapabilityPlan`.
2. ✅ **Defensive nil check removed** — `if req == nil` guard eliminated from `collectContainerNames`.
3. ✅ **Config parsing tests added** — table-driven `TestParseCapabilityImplementations` covers empty string, empty map, valid rootless/privileged, invalid implementation, unknown key, and malformed JSON.
4. ✅ **Missing workload test cases added** — `TestStartWorkloadRejectsDockerCapabilityNotConfigured`, `TestStartWorkloadRejectsDockerContainerNameConflict`, `TestStartWorkloadRejectsDockerVolumeNameConflict` all verify correct gRPC status codes and error messages.

No new issues introduced. LGTM.
